crypto/pkcs7/bio/cipher.c

```diff
@@ -0,0 +1,310 @@
+    if (bytes_written <= 0) {
+      BIO_copy_next_retry(b);
+      break;
+    }
+    ctx->buf_off += bytes_written;
+    ctx->buf_len -= bytes_written;
+  }
+  return bytes_consumed;
+}
+
+static long enc_ctrl(BIO *b, int cmd, long num, void *ptr) {
+  GUARD_PTR(b);
+  long ret = 1;
+
+  BIO_ENC_CTX *ctx = BIO_get_data(b);
+  BIO *next = BIO_next(b);
+  if (ctx == NULL) {
+    return 0;
+  }
+
+  switch (cmd) {
+    case BIO_CTRL_RESET:
+      ctx->done = 0;
+      ctx->ok = 1;
+      ctx->buf_off = 0;
+      ctx->buf_len = 0;
+      OPENSSL_cleanse(ctx->buf, sizeof(ctx->buf));
+      if (!EVP_CipherInit_ex(ctx->cipher, NULL, NULL, NULL, NULL,
+                             EVP_CIPHER_CTX_encrypting(ctx->cipher))) {
+        return 0;
+      }
+      ret = BIO_ctrl(next, cmd, num, ptr);
+      break;
+    case BIO_CTRL_EOF:
+      if (ctx->done) {
+        ret = 1;
+      } else {
+        ret = BIO_ctrl(next, cmd, num, ptr);
+      }
+      break;
+    case BIO_CTRL_WPENDING:
+    case BIO_CTRL_PENDING:
+      // Return number of bytes left to process if we have anything buffered,
+      // else consult underlying BIO.
+      ret = ctx->buf_len;
+      if (ret <= 0) {
+        ret = BIO_ctrl(next, cmd, num, ptr);
+      }
+      break;
+    case BIO_CTRL_FLUSH:
+      ret = enc_flush(b, next, ctx);
+      if (ret <= 0) {
+        break;
+      }
+      // Flush the underlying BIO
+      ret = BIO_ctrl(next, cmd, num, ptr);
+      BIO_copy_next_retry(b);
+      break;
+    case BIO_C_GET_CIPHER_STATUS:
+      ret = (long)ctx->ok;
+      break;
+    // OpenSSL implements these, but because we don't need them and cipher BIO
+    // is internal, we can fail loudly if they're called. If this case is hit,
+    // it likely means you're making a change that will require implementing
+    // these.
+    case BIO_CTRL_DUP:
+    case BIO_CTRL_GET_CALLBACK:
+    case BIO_CTRL_SET_CALLBACK:
+    case BIO_C_DO_STATE_MACHINE:
+    case BIO_C_GET_CIPHER_CTX:
+      OPENSSL_PUT_ERROR(PKCS7, ERR_R_BIO_LIB);
+      return 0;
+    default:
+      ret = BIO_ctrl(next, cmd, num, ptr);
+      break;
+  }
+  return ret;
+}
+
+int BIO_set_cipher(BIO *b, const EVP_CIPHER *c, const unsigned char *key,
+                   const unsigned char *iv, int enc) {
+  GUARD_PTR(b);
+  GUARD_PTR(c);
+
+  BIO_ENC_CTX *ctx = BIO_get_data(b);
+  if (ctx == NULL) {
+    return 0;
+  }
+
+  // We only support a modern subset of available EVP_CIPHERs. Other ciphers
+  // (e.g. DES) and cipher modes (e.g. CBC, CCM) had issues with block alignment
+  // and padding during testing, so they're forbidden for now.
+  const EVP_CIPHER *kSupportedCiphers[] = {
+      EVP_aes_128_cbc(),       EVP_aes_128_ctr(), EVP_aes_128_ofb(),
+      EVP_aes_256_cbc(),       EVP_aes_256_ctr(), EVP_aes_256_ofb(),
+      EVP_chacha20_poly1305(),
+  };
+  int supported = 0;
+  for (size_t i = 0; i < sizeof(kSupportedCiphers) / sizeof(EVP_CIPHER *);
```

Comment:
NP: Compute the size of the array into its own constant.
```
  const size_t kSupportedCiphersCount =
      sizeof(kSupportedCiphers) / sizeof(EVP_CIPHER *);
  int supported = 0;
  for (size_t i = 0; i < kSupportedCiphersCount; i++) {
```

Reply:
Yeah, the concern for me is not performance so much as clarity/readability of the code.